src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java

```diff
@@ -242,13 +236,25 @@
                 });
     }
 
-    public static void saveMetricsRDD(final JavaPairRDD<String, DuplicationMetrics> metricsRDD, final String metricsOutputPath, AuthHolder authHolder) {
-        final MetricsFile<DuplicationMetrics, Double> metrics = new MetricsFile<>();
-        for (final Map.Entry<String, DuplicationMetrics> entry : metricsRDD.collectAsMap().entrySet()) {
-            metrics.addMetric(entry.getValue());
+    public static void saveMetricsRDD(final MetricsFile<DuplicationMetrics, Double> result, final JavaPairRDD<String, DuplicationMetrics> metricsRDD, final LibraryIdGenerator libraryIdGenerator, final String metricsOutputPath, AuthHolder authHolder) {
```

Comment:
Could you make this take a SamFileHeader and make the LibraryIdGenerator internally?

Add a javadoc explaining that `result` may be initialized with headers, and that the SamFileHeader is needed in order to include libraries that didn't have any duplicates.